stdlib/json.gr

```diff
@@ -2069,3 +2069,298 @@ provide let parse: (str: String) => Result<Json, JsonParseError> = (str: String)
+          [] => List.reverse([(propertyName, newValue), ...acc]),
+        }
+      }
+
+      match (json) {
+        JsonObject(props) => Some(JsonObject(replaceFirst([], props))),
+        _ => None,
+      }
+    } }
+
+  /**
+   * Wraps a lens to permit nullable values in addition to the original value
+   * type of the given lens. During a `get` operation if the lens matches then
+   * the result will be enclosed in `Some`; if the lens does not match but the
+   * value focused is null, then the lens will still successfully match and
+   * `None` will be returned.
+   * 
+   * @example assert get(nullable(number), JsonNumber(123)) == Some(Some(123))
+   * @example assert get(nullable(number), JsonNull) == Some(None)
+   * @example assert get(nullable(number), JsonString("abc")) == None
+   * @example assert set(nullable(number), Some(123), JsonString("abc")) == Some(JsonNumber(123))
+   * 
+   * @since v0.7.0
+   */
+  provide let nullable = lens =>
+    { get: json => {
+      match (get(lens, json)) {
+        Some(x) => Some(Some(x)),
+        None => {
+          match (json) {
+            JsonNull => Some(None),
+            _ => {
+              match (get(lens, json)) {
+                Some(x) => Some(Some(x)),
+                None => None,
+              }
+            },
+          }
+        },
+      }
+    }, set: (newValue, json) => {
+      match (newValue) {
+        Some(x) => lens.set(x, json),
+        None => Some(JsonNull),
+      }
+    } }
+
+  /**
+   * Reverse lens composition.
+   * 
+   * @param lens1: The lens which will be applied first
+   * @param lens2: The lens which will be applied second
+   * @returns A lens which combines the two given lenses, passing through the first and then the second
+   * 
+   * @example assert get(prop("x") ||> number, JsonObject([("x", JsonNumber(123))])) == Some(123)
+   * @example
+   * assert set(prop("x") ||> string, "new", JsonObject([("x", JsonNumber(123))])) ==
+   *   Some(JsonObject([("x", JsonString("new"))]))
+   * 
+   * @since v0.7.0
+   */
+  provide let (||>) = (lens1, lens2) =>
+    { get: json => {
+      match (lens1.get(json)) {
+        Some(x) => lens2.get(x),
+        None => None,
+      }
+    }, set: (newValue, target) => {
+      match (lens1.get(target)) {
+        Some(x) => match (lens2.set(newValue, x)) {
+          Some(y) => lens1.set(y, target),
+          None => None,
+        },
+        None => None,
+      }
+    } }
+
+  /**
+   * Creates a lens whose focus is a given property path within a JSON object tree.
+   * 
+   * @param propertyNames: The property path of the JSON object to create a focus on
+   * @returns A lens whose focus is the given property path of a JSON object
+   * 
+   * @example
+   * let nestedObj = JsonObject([("a", JsonObject([("b", JsonNumber(123))]))])
+   * assert get(propPath(["a", "b"]), nestedObj) == Some(JsonNumber(123))
+   * 
+   * @since v0.7.0
+   */
+  provide let propPath = propertyNames => {
```

Comment:
can we move this with the other property stuff, I feel like putting it all the way down here isn't going to make it easy to find.

Reply:
Yeah initially did this so that I could use the `||>` function but I'll just use the alternative provide syntax to change the ordering in the docs